src/state_machine.rs

```diff
@@ -400,6 +410,13 @@ impl StateMachine {
                 };
                 self.state = new_state;
             }
+            IntoNode { gen_pfx_info } => {
+                let new_state = match mem::replace(&mut self.state, State::Terminated) {
```

Comment:
I don't think there is a way to do it without the temporary state and without involving `unsafe`, but if you think of one, I'm all ears :D

Reply:
It also follow the existing pattern.

Reply:
I'd love to have a proper solution to this problem! Maybe it would be slightly cleaner to isolate the place where we assign the temporary value, to separate it from the business logic itself (that's what I did [elsewhere](https://github.com/poanetwork/hbbft/blob/a64f62506d1c3e98ad805c61f2c825439d2025d3/src/subset/proposal_state.rs#L162)); so we'd have an
```rust
fn do_apply_transition(state: State, crust_tx: …, …, transition: …, outbox: …) -> State;
```
and in `apply_transition` itself we'd only do the swap:
```rust
let state = mem::replace(&mut self.state, State::Terminated);
let new_state = do_apply_transition(state, &mut self crust_tx, …, transition, outbox);
self.state = new_state;
```

Reply:
Adding a function like that should be doable. Not sure we can do much better than that though.